src/common/stats/StatsManager.cpp

```diff
@@ -311,6 +323,153 @@ StatusOr<StatsManager::VT> StatsManager::readHisto(const std::string& counterNam
+        range = TimeRange::ONE_MINUTE;
+    } else if (parts[2] == "600") {
+        range = TimeRange::TEN_MINUTES;
+    } else if (parts[2] == "3600") {
+        range = TimeRange::ONE_HOUR;
+    } else {
+        // Unsupported time range
+        std::string err = folly::stringPrintf("Unsupported time range \"%s\"", parts[2].c_str());
+        return Status::Error(std::move(err));
+    }
+
+    StatsMethod method;
+    folly::toLowerAscii(parts[1]);
+    if (parts[1] == "sum") {
+        method = StatsMethod::SUM;
+    } else if (parts[1] == "count") {
+        method = StatsMethod::COUNT;
+    } else if (parts[1] == "avg") {
+        method = StatsMethod::AVG;
+    } else if (parts[1] == "rate") {
+        method = StatsMethod::RATE;
+    // } else if (parts[1][0] == 'p') {   // TODO(shylock)
+    } else {
+        std::string err = folly::stringPrintf(
+            "Unsupported statistic method \"%s\"", parts[1].c_str());
+        LOG(ERROR) << err;
+        return Status::Error(std::move(err));
+    }
+
+    return StatsManager::ParsedName {
+        name,
+        method,
+        range,
+    };
+}
+
+
+// The raw metric format based on JSON
+// 1. Gauge, time serial value
+// 2. Histogram, time serial value distribution
+// The raw data can be transformed to multiple specified foramt (various user-defined metric format)
+// E.G.
+// {
+//     "name": "meta",
+//     "gauges": [...],
+//     "histograms:" [...],
+// }
+// Gauge
+// {
+//     "name": "xxxxx",
+//     "value": 33,
+//     "labels": [
+//         {"name": "name", "value": "nebula"},
+//         {"name": "type", "value": "qps"}
+//     ]
+// }
+// Histogram
+// {
+//     "name": "xxxxx",
+//     "value_range": [0, 100],
+//     "buckets": [2, 3, 0, 11, ...],
+//     "sum": 233,
+//     "count": 332,
+//     "labels": [
+//         {"name": "name", "value": "nebula"},
+//         {"name": "type", "value": "latency"}
+//     ]
+// }
+//
+/*static*/ std::string StatsManager::toJson() {
+    // TODO(shylock) There too much hard code but could fit current usage
+    // Consider defined a Format to Create/Evaluate metric
+    // Draft, https://www.yuque.com/docs/share/c2ec6b5a-77a2-47d6-9676-4c4b1b4325e4?#
+    constexpr char kGauges[] = "gauges";
+    constexpr char kHistograms[] = "histograms";
+    auto& sm = get();
+    folly::dynamic obj = folly::dynamic::object(kGauges, folly::dynamic::array())
+        (kHistograms, folly::dynamic::array());
+    folly::dynamic labels = folly::dynamic::array(
+        folly::dynamic(folly::dynamic::object("name", "root")("value", "nebula")));
+
+    // insert
+    for (auto& index : sm.nameMap_) {
+        auto parsedName = sm.idParsedName_[index.second].get();
+        std::string name = parsedName != nullptr ? parsedName->name : index.first;
+        if (StatsManager::isStatIndex(index.second)) {
+            auto result = StatsManager::readStats(index.second,
+                    StatsManager::TimeRange::ONE_MINUTE,
```

Comment:
The `Stats` uses one minute, but the  `Histo` uses one hour, why? and the `Stats` only get the avg, why, what about other metrics?

Reply:
> The `Stats` uses one minute, but the `Histo` uses one hour, why? and the `Stats` only get the avg, why, what about other metrics?

It's hard code. Consider enable user to define specified metrics like https://www.yuque.com/docs/share/c2ec6b5a-77a2-47d6-9676-4c4b1b4325e4?#